railgun/internal/proxy/clientgo_cached_proxy_resolver.go

```diff
@@ -0,0 +1,227 @@
+	for {
+		select {
+		case cobj := <-p.update:
+			if err := p.cacheUpdate(cobj); err != nil {
+				p.logger.Error(err, "object could not be updated in the cache and will be discarded")
+				break
+			}
+			updates = true
+		case cobj := <-p.del:
+			if err := p.cacheDelete(cobj); err != nil {
+				p.logger.Error(err, "object could not be deleted from the cache and will be discarded")
+				break
+			}
+			updates = true
+		case <-syncTicker.C:
+			// if there are no relevant object updates for this cycle, then there's no reason to
+			// bother the Kong Proxy with updates from the cache.
+			if !updates {
+				break
+			}
+			if err := p.updateKongAdmin(); err != nil {
+				p.logger.Error(err, "could not update kong admin")
+			}
+			updates = false
+		case <-p.ctx.Done():
+			p.logger.Info("the proxy cache server's context is done, shutting down")
+			return
+		}
+	}
+}
+
+// -----------------------------------------------------------------------------
+// Client Go Cached Proxy Resolver - Private Methods - Cache Server Utils
+// -----------------------------------------------------------------------------
+
+// objectKey provides a key unique to the cacheServer for tracking objects that are being resolved.
+func (p *clientgoCachedProxyResolver) clientObjectKey(obj client.Object) string {
+	return fmt.Sprintf("%s/%s/%s", obj.GetObjectKind().GroupVersionKind().String(), obj.GetNamespace(), obj.GetName())
+}
+
+// cacheUpdate caches the provided object to the proxy cache and the provided object tracker and reports any errors.
+func (p *clientgoCachedProxyResolver) cacheUpdate(cobj *cachedObject) error {
+	cobj.err = p.cache.Add(cobj.runtimeObj)
+	return cobj.err
+}
+
+// cacheDelete removes the cache entry the provided object from the proxy cache and the provided object tracker and reports any errors.
+func (p *clientgoCachedProxyResolver) cacheDelete(cobj *cachedObject) error {
+	cobj.err = p.cache.Delete(cobj.runtimeObj)
+	return cobj.err
+}
+
+// updateKongAdmin will take whatever the current state of the Proxy.cache is an convert that to Kong DSL
+// and apply the resulting configuration to the Kong Admin API.
+func (p *clientgoCachedProxyResolver) updateKongAdmin() error {
+	// build the kongstate object from the Kubernetes objects in the storer
+	storer := store.New(*p.cache, p.ingressClassName, p.processClasslessIngressV1, p.processClasslessIngressV1Beta1, p.processClasslessKongConsumer, logrus.StandardLogger())
+	kongstate, err := parser.Build(logrus.StandardLogger(), storer)
+	if err != nil {
+		return err
+	}
+
+	// generate the deck configuration to be applied to the admin API
+	targetConfig := deckgen.ToDeckContent(p.ctx, logrus.StandardLogger(), kongstate, nil, nil)
+
+	// apply the configuration update in Kong
+	timedCtx, cancel := context.WithTimeout(p.ctx, 10*time.Second)
+	defer cancel()
+	_, err = sendconfig.PerformUpdate(timedCtx, logrus.StandardLogger(), &p.kongConfig, true, false, targetConfig, nil, nil, nil)
```

Comment:
Would it be possible to maintain the old behavior here, where we do store the SHA returned by sendconfig and then send it on subsequent requests, so that nil updates are ignored?

On `clientgoCachedProxyResolver`, we have:

> This object's attributes are immutable (private), and it is threadsafe.

Which suggests we don't want the old approach of [stuffing the SHA into the KongController state](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.2.0/internal/ingress/controller/kong.go#L71). Maybe there's somewhere else we can store it?

My concern is that excess updates have been a performance concern in the past, and this seems like it'd exacerbate them without #1259, which looks like it'd require a decent amount of new work. If we can maintain the config SHA-based means of preventing config updates from irrelevant events, doing so seems prudent.

Reply:
I think you're right and that we can use the config SHA. I've created this follow up which links back to this comment:

https://github.com/Kong/kubernetes-ingress-controller/issues/1280

If you're OK with it, since this PR has already gotten fairly large I would like to make this improvement as a follow up, but it can be the next thing I work on and a blocker for alpha1.